compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,409 @@
+        // is true.
+        ty::Adt(adt_def, ..) => {
+            // S<length><name>
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.opt_item_name(def_id).unwrap());
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                s.push('v');
+            } else {
+                s.push_str(&format!("{}", name.chars().count()));
+                s.push_str(&name);
+                compress(dict, ty, &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // S<length><name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.opt_item_name(*def_id).unwrap());
+            s.push_str(&format!("{}", name.chars().count()));
+            s.push_str(&name);
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E
+            let mut s = String::from("u2fn");
+            let name = format!("{}", tcx.opt_item_name(*def_id).unwrap());
+            s.push_str(&format!("{}", name.chars().count()));
+            s.push_str(&name);
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Closure(_, substs) | ty::Generator(_, substs, _) => {
+            // u7closureI<return-type><element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u7closureI");
+            let sig = substs.as_closure().sig();
+            // Push str without the initial 'F' but with the closing 'E'
+            s.push_str(&(encode_fnsig(tcx, &sig.skip_binder(), dict)[1..]));
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Pointer types
+        ty::Ref(_, ty0, mutbl) => {
+            // [U3mut]u3refI<element-type>E as vendor extended type qualifier and type
+            let mut s = String::from(match mutbl {
+                hir::Mutability::Not => "u3refI",
+                hir::Mutability::Mut => "U3mutu3refI",
+            });
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types));
+            s.push('E');
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::RawPtr(tm) => {
+            // P[K]<element-type>
+            let mut s = String::from(match tm.mutbl {
+                hir::Mutability::Not => "PK",
+                hir::Mutability::Mut => "P",
+            });
+            s.push_str(&encode_ty(tcx, tm.ty, dict, encode_c_types));
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
```

Comment:
The production used here are:

```
  <type> ::= <qualified-type>
         ::= P <type>        # pointer
         ...
         ::= <substitution>  # See Compression below
```

The `<type> ::= <qualified-type>` should create a separate substitution.

For example `pub extern "C" fn f(a: *const i32, b: *const *const i32) {}` is currently mangled as `_ZTSFvPKu3i32PKS0_E` which demangles to `typeinfo name for void (i32 const*, i32 const*)`.

Reply:
If `S_ = u3i32` and `S0_ = PKu3i32`, shouldn't `PKS0_` in `_ZTSFvPKu3i32PKS0_E` demangle to `const* const* i32` (i.e., `PKPKu3i32`)?

Reply:
In between `u3i32` and `PKu3i32` there is `Ku3i32` which is also added to the dictionary, so `S0_ = Ku3i32`. The dictionary key will need to be something more than just a `Ty`.

```console
$ c++filt
_ZTSFvPKu3i32S_E
typeinfo name for void (i32 const*, i32)
_ZTSFvPKu3i32S0_E
typeinfo name for void (i32 const*, i32 const)
_ZTSFvPKu3i32S1_E   
typeinfo name for void (i32 const*, i32 const*)
```

Reply:
I see. Thanks for pointing it out! I'll fix it. I believe it'd also be the case for `ty:Ref`, right?

Reply:
Done.